File: src/cli/main/run-cli/resolve-output-path/derive-output-basename.test.ts
import { describe, expect, test } from "vitest";

import type { ParsedRootQuery } from "../../../../root-query/parsed-root-query.js";
import { ROOT_QUERY_KIND } from "../../../../root-query/root-query-kind.js";
import { deriveOutputBasename } from "./derive-output-basename.js";

const noRadius = {
  descendants: null,
  ancestors: null,
  context: null,
} satisfies Record<"descendants" | "ancestors" | "context", number | null>;

const name = (n: string): ParsedRootQuery => ({
  kind: ROOT_QUERY_KIND.Name,
  name: n,
  raw: n,
});

const line = (n: number): ParsedRootQuery => ({
  kind: ROOT_QUERY_KIND.Line,
  line: n,
  raw: String(n),
});

const lineName = (n: number, id: string): ParsedRootQuery => ({
  kind: ROOT_QUERY_KIND.LineName,
  line: n,
  name: id,
  raw: `${n}:${id}`,
});

const range = (s: number, e: number): ParsedRootQuery => ({
  kind: ROOT_QUERY_KIND.Range,
  start: s,
  end: e,
  raw: `${s}-${e}`,
});

const rangeName = (s: number, e: number, id: string): ParsedRootQuery => ({
  kind: ROOT_QUERY_KIND.RangeName,
  start: s,
  end: e,
  name: id,
  raw: `${s}-${e}:${id}`,
});

describe("deriveOutputBasename: root tokenization", () => {
  test("kind=name uses the identifier", () => {
    const actual = deriveOutputBasename({
      roots: [name("value")],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("value");
  });

  test("kind=line is l<n>", () => {
    const actual = deriveOutputBasename({
      roots: [line(42)],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("l42");
  });

  test("kind=line-name is l<n>-<id> with single hyphen", () => {
    const actual = deriveOutputBasename({
      roots: [lineName(42, "render")],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("l42-render");
  });

  test("kind=range is l<n>-<m>", () => {
    const actual = deriveOutputBasename({
      roots: [range(10, 12)],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("l10-12");
  });

  test("kind=range-name is l<n>-<m>-<id>", () => {
    const actual = deriveOutputBasename({
      roots: [rangeName(10, 12, "render")],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("l10-12-render");
  });

  test("multiple roots are joined with +", () => {
    const actual = deriveOutputBasename({
      roots: [name("value"), name("foo")],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("value+foo");
  });

  test("multiple roots mixing kinds", () => {
    const actual = deriveOutputBasename({
      roots: [lineName(42, "render"), name("foo")],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("l42-render+foo");
  });
});

describe("deriveOutputBasename: radius suffix", () => {
  test("no radius flag → no suffix", () => {
    const actual = deriveOutputBasename({
      roots: [name("value")],
      ...noRadius,
      inputPath: "",
    });

    expect(actual).toBe("value");
  });

  test("-A only → -a<N>", () => {
    const actual = deriveOutputBasename({
      roots: [name("value")],
      descendants: 1,
      ancestors: null,
      context: null,
      inputPath: "",
    });

    expect(actual).toBe("value-a1");
  });

  test("-B only → -b<N>", () => {
    const actual = deriveOutputBasename({
      roots: [name("param")],
      descendants: null,
      ancestors: 2,
      context: null,
      inputPath: "",
    });

    expect(actual).toBe("param-b2");
  });

  test("-C only → -c<N>", () => {
    const actual = deriveOutputBasename({
      roots: [range(10, 12)],
      descendants: null,
      ancestors: null,
      context: 2,
      inputPath: "",
    });

    expect(actual).toBe("l10-12-c2");
  });

  test("-A and -B → -a<N>-b<M>", () => {
    const actual = deriveOutputBasename({
      roots: [name("v")],
      descendants: 1,
      ancestors: 2,
      context: null,
      inputPath: "",
    });

    expect(actual).toBe("v-a1-b2");
  });

  test("-B and -C → alphabetical -b<N>-c<M>", () => {
    const actual = deriveOutputBasename({
      roots: [name("v")],
      descendants: null,
      ancestors: 2,
      context: 3,
      inputPath: "",
    });

    expect(actual).toBe("v-b2-c3");
  });

  test("-C and -A → alphabetical -a<N>-c<M>", () => {
    const actual = deriveOutputBasename({
      roots: [name("v")],
      descendants: 7,
      ancestors: null,
      context: 3,
      inputPath: "",
    });

    expect(actual).toBe("v-a7-c3");
  });

  test("-A and -B together drop -C from the filename (C is then redundant)", () => {
    const actual = deriveOutputBasename({
      roots: [name("v")],
      descendants: 1,
      ancestors: 2,
      context: 3,
      inputPath: "",
    });

    expect(actual).toBe("v-a1-b2");
  });
});

describe("deriveOutputBasename: input file fallback", () => {
  test("strips .ts extension", () => {
    const actual = deriveOutputBasename({
      roots: [],
      ...noRadius,
      inputPath: "foo.ts",
    });

    expect(actual).toBe("foo");
  });

  test("preserves camelCase basename", () => {
    const actual = deriveOutputBasename({
      roots: [],
      ...noRadius,
      inputPath: "fooBar.ts",
    });

    expect(actual).toBe("fooBar");
  });

  test("preserves kebab-case basename", () => {
    const actual = deriveOutputBasename({
      roots: [],
      ...noRadius,
      inputPath: "foo-bar.ts",
    });

    expect(actual).toBe("foo-bar");
  });

  test("strips .tsx", () => {
    const actual = deriveOutputBasename({
      roots: [],
      ...noRadius,
      inputPath: "Component.tsx",
    });

    expect(actual).toBe("Component");
  });

  test("strips only the last extension", () => {
    const actual = deriveOutputBasename({
      roots: [],
      ...noRadius,
      inputPath: "foo.test.ts",
    });

    expect(actual).toBe("foo.test");
  });

  test("no extension → keeps full basename", () => {
    const actual = deriveOutputBasename({
      roots: [],
      ...noRadius,
      inputPath: "Makefile",
    });

    expect(actual).toBe("Makefile");
  });

  test("strips path components", () => {
    const actual = deriveOutputBasename({
      roots: [],
      ...noRadius,
      inputPath: "src/deep/foo.ts",
    });

    expect(actual).toBe("foo");
  });
});

File: crates/unsnarl/src/cli/run_cli/resolve_output_path/derive_output_basename_test.rs
use std::path::Path;

use super::*;

use unsnarl_ir::SourceLine;
use unsnarl_root_query::GenerationCount;

fn name(n: &str) -> ParsedRootQuery {
    ParsedRootQuery::Name {
        name: n.to_string(),
        raw: n.to_string(),
    }
}

fn line(n: u32) -> ParsedRootQuery {
    ParsedRootQuery::Line {
        line: SourceLine(n),
        raw: n.to_string(),
    }
}

fn line_name(n: u32, id: &str) -> ParsedRootQuery {
    ParsedRootQuery::LineName {
        line: SourceLine(n),
        name: id.to_string(),
        raw: format!("{n}:{id}"),
    }
}

fn range(s: u32, e: u32) -> ParsedRootQuery {
    ParsedRootQuery::Range {
        start: SourceLine(s),
        end: SourceLine(e),
        raw: format!("{s}-{e}"),
    }
}

fn range_name(s: u32, e: u32, id: &str) -> ParsedRootQuery {
    ParsedRootQuery::RangeName {
        start: SourceLine(s),
        end: SourceLine(e),
        name: id.to_string(),
        raw: format!("{s}-{e}:{id}"),
    }
}

mod root_tokenization {
    use super::*;

    #[test]
    fn kind_name_uses_identifier() {
        let actual = derive_output_basename(&[name("value")], None, None, None, Path::new(""));
        assert_eq!(actual, "value");
    }

    #[test]
    fn kind_line_is_l_n() {
        let actual = derive_output_basename(&[line(42)], None, None, None, Path::new(""));
        assert_eq!(actual, "l42");
    }

    #[test]
    fn kind_line_name_is_l_n_dash_id_with_single_hyphen() {
        let actual =
            derive_output_basename(&[line_name(42, "render")], None, None, None, Path::new(""));
        assert_eq!(actual, "l42-render");
    }

    #[test]
    fn kind_range_is_l_n_dash_m() {
        let actual = derive_output_basename(&[range(10, 12)], None, None, None, Path::new(""));
        assert_eq!(actual, "l10-12");
    }

    #[test]
    fn kind_range_name_is_l_n_dash_m_dash_id() {
        let actual = derive_output_basename(
            &[range_name(10, 12, "render")],
            None,
            None,
            None,
            Path::new(""),
        );
        assert_eq!(actual, "l10-12-render");
    }

    #[test]
    fn multiple_roots_joined_with_plus() {
        let actual = derive_output_basename(
            &[name("value"), name("foo")],
            None,
            None,
            None,
            Path::new(""),
        );
        assert_eq!(actual, "value+foo");
    }

    #[test]
    fn multiple_roots_mixing_kinds() {
        let actual = derive_output_basename(
            &[line_name(42, "render"), name("foo")],
            None,
            None,
            None,
            Path::new(""),
        );
        assert_eq!(actual, "l42-render+foo");
    }
}

mod radius_suffix_inclusion {
    use super::*;

    #[test]
    fn no_radius_flag_yields_no_suffix() {
        let actual = derive_output_basename(&[name("value")], None, None, None, Path::new(""));
        assert_eq!(actual, "value");
    }

    #[test]
    fn descendants_only_yields_a_n() {
        let actual = derive_output_basename(
            &[name("value")],
            Some(GenerationCount(1)),
            None,
            None,
            Path::new(""),
        );
        assert_eq!(actual, "value-a1");
    }

    #[test]
    fn ancestors_only_yields_b_n() {
        let actual = derive_output_basename(
            &[name("param")],
            None,
            Some(GenerationCount(2)),
            None,
            Path::new(""),
        );
        assert_eq!(actual, "param-b2");
    }

    #[test]
    fn context_only_yields_c_n() {
        let actual = derive_output_basename(
            &[range(10, 12)],
            None,
            None,
            Some(GenerationCount(2)),
            Path::new(""),
        );
        assert_eq!(actual, "l10-12-c2");
    }

    #[test]
    fn a_and_b_yields_a_n_b_m() {
        let actual = derive_output_basename(
            &[name("v")],
            Some(GenerationCount(1)),
            Some(GenerationCount(2)),
            None,
            Path::new(""),
        );
        assert_eq!(actual, "v-a1-b2");
    }

    #[test]
    fn b_and_c_yields_alphabetical_b_n_c_m() {
        let actual = derive_output_basename(
            &[name("v")],
            None,
            Some(GenerationCount(2)),
            Some(GenerationCount(3)),
            Path::new(""),
        );
        assert_eq!(actual, "v-b2-c3");
    }

    #[test]
    fn c_and_a_yields_alphabetical_a_n_c_m() {
        let actual = derive_output_basename(
            &[name("v")],
            Some(GenerationCount(7)),
            None,
            Some(GenerationCount(3)),
            Path::new(""),
        );
        assert_eq!(actual, "v-a7-c3");
    }

    #[test]
    fn a_and_b_drop_c_from_the_filename() {
        let actual = derive_output_basename(
            &[name("v")],
            Some(GenerationCount(1)),
            Some(GenerationCount(2)),
            Some(GenerationCount(3)),
            Path::new(""),
        );
        assert_eq!(actual, "v-a1-b2");
    }
}

mod input_file_fallback {
    use super::*;

    #[test]
    fn strips_ts_extension() {
        let actual = derive_output_basename(&[], None, None, None, Path::new("foo.ts"));
        assert_eq!(actual, "foo");
    }

    #[test]
    fn preserves_camel_case_basename() {
        let actual = derive_output_basename(&[], None, None, None, Path::new("fooBar.ts"));
        assert_eq!(actual, "fooBar");
    }

    #[test]
    fn preserves_kebab_case_basename() {
        let actual = derive_output_basename(&[], None, None, None, Path::new("foo-bar.ts"));
        assert_eq!(actual, "foo-bar");
    }

    #[test]
    fn strips_tsx_extension() {
        let actual = derive_output_basename(&[], None, None, None, Path::new("Component.tsx"));
        assert_eq!(actual, "Component");
    }

    #[test]
    fn strips_only_the_last_extension() {
        let actual = derive_output_basename(&[], None, None, None, Path::new("foo.test.ts"));
        assert_eq!(actual, "foo.test");
    }

    #[test]
    fn no_extension_keeps_full_basename() {
        let actual = derive_output_basename(&[], None, None, None, Path::new("Makefile"));
        assert_eq!(actual, "Makefile");
    }

    #[test]
    fn strips_path_components() {
        let actual = derive_output_basename(&[], None, None, None, Path::new("src/deep/foo.ts"));
        assert_eq!(actual, "foo");
    }
}
